hivemind/utils/grpc.py

```diff
@@ -58,6 +72,10 @@ def deserialize_torch_tensor(serialized_tensor: runtime_pb2.Tensor) -> torch.Ten
         stds = torch.as_tensor(np.frombuffer(stds, dtype=np.float32)).view(*stats_size)
         array = np.frombuffer(serialized_tensor.buffer[:-8 * stats_count], dtype=np.float16)
         tensor = torch.as_tensor(array).to(torch.float32).view(*serialized_tensor.size).mul_(stds).add_(means)
+    elif serialized_tensor.compression == CompressionType.FLOAT16:
+        array = np.frombuffer(serialized_tensor.buffer, dtype=np.float16).copy()
+        tensor = torch.as_tensor(array).view(*serialized_tensor.size).\
+            requires_grad_(serialized_tensor.requires_grad).to(torch.float32)
```

Comment:
```suggestion
        tensor = torch.as_tensor(array).view(*serialized_tensor.size)\
            .to(torch.float32).requires_grad_(serialized_tensor.requires_grad)
```
Please cast to float32 first and set requires_grad second - this will avoid an unnecessary cast of gradients to fp32. Otherwise everything's great!